apps/agreement/contracts/disputable/Delay.sol

```diff
@@ -0,0 +1,459 @@
+        uint64 executableAt;            // Date time when the delayable can be executed
+        DelayableState state;           // Current state of the delayable
+        uint256 actionId;               // Identification number of the action in the context of the agreement
+        bytes script;                   // Action script to be executed
+    }
+
+    struct TokenBalancePermission {
+        ERC20 token;                    // ERC20 token to be used for custom permissions based on token balance
+        uint256 balance;                // Amount of tokens used for custom permissions
+    }
+
+    /**
+    * @dev Auth modifier restricting access only for addresses that can submit actions
+    */
+    modifier onlySubmitter() {
+        require(_canSubmit(msg.sender), ERROR_AUTH_FAILED);
+        _;
+    }
+
+    uint64 public delayPeriod;
+    Delayable[] private delayables;
+
+    TokenBalancePermission private submitTokenBalancePermission;
+    TokenBalancePermission private challengeTokenBalancePermission;
+
+    /**
+    * @notice Initialize Delay app with `@transformTime(_delayPeriod)`, linked to the Agreement `_agreement` with the following requirements:
+    * @notice - `@tokenAmount(_collateralToken, _actionCollateral)` collateral for submitting actions
+    * @notice - `@tokenAmount(_collateralToken, _challengeCollateral)` collateral for challenging actions
+    * @notice - `@transformTime(_challengeDuration)` for the challenge duration
+    * @notice - Submit permission: `_submitPermissionToken == 0 ? 'None' : @tokenAmount(_submitPermissionToken, _submitPermissionBalance)`
+    * @notice - Challenge per: `_challengePermissionBalance == 0 ? 'None' : @tokenAmount(_challengePermissionToken, _challengePermissionBalance)`
+    * @param _delayPeriod Duration in seconds during which an action is delayed before being executed
+    * @param _agreement Agreement instance to be linked
+    * @param _collateralToken Address of the ERC20 token to be used for collateral
+    * @param _actionCollateral Amount of collateral tokens that will be locked every time an delayable is submitted
+    * @param _challengeCollateral Amount of collateral tokens that will be locked every time an delayable is challenged
+    * @param _challengeDuration Duration in seconds of the challenge, during this time window the submitter can answer the challenge
+    * @param _submitPermissionToken ERC20 token to be used for custom submitting permissions based on token balance
+    * @param _submitPermissionBalance Amount of `_submitPermissionToken` tokens for custom signing permissions
+    * @param _challengePermissionToken ERC20 token to be used for custom challenge permissions based on token balance
+    * @param _challengePermissionBalance Amount of `_challengePermissionBalance` tokens for custom challenge permissions
+    */
+    function initialize(
+        uint64 _delayPeriod,
+        IAgreement _agreement,
```

Comment:
I am really not sure about requiring the Agreement during initialization and not being changeable.

It makes it more difficult to install in existing organizations who may already have these apps installed, and harder to transition the app if they'd like to uninstall or migrate to another Agreement.